Add implementation of class related features to DG (#176)
AY2223S1-CS2103T-T12-1/tp: docs/DeveloperGuide.md

## **Implementation**

### Creating and deleting classes

Class information is stored as `ModuleClass` objects, which captures the name of the class as well as the sessions created for the class. When the user creates a class, the program creates a new `ModuleClass` object with the class name and adds it to the collection of `ModuleClass` objects previously created, which is managed by the `TaAssist` class.

> When deleting classes, all students previously assigned to those classes will have the classes unassigned from them, before the `ModuleClass` objects corresponding to those classes are removed from the collection captured by `TaAssist`.

Reviewer: I think `module classes` will sound less confusing than `classes` cos `classes` can mean Java classes. Same for other parts of the DG.

Author: Alright! We shall keep it consistent as module class as well in DG.

The following methods in `TaAssist` manages the adding and deleting of classes from the collection:

- `TaAssist#addModuleClass(ModuleClass moduleClass)` - Adds the provided class to the list of classes created.
- `TaAssist#removeModuleClass(ModuleClass moduleClass)` - Removes the provided class from the list of classes created.

### Assigning students to a class
<img src="images/AssignCommandSequenceDiagram.png" width="700" />

> Each student object contains a collection of `StudentModuleData` where classes and the grades the student obtained for the sessions of the classes are stored. When the user tries to assign students to a class, we are essentially creating a new `StudentModuleData` object to be added to the collection for each student.

Reviewer: ```suggestion
Each student object contains a collection of `StudentModuleData` where classes and the grades the student obtained for the sessions of the classes are stored. When the user assigns students to a class, a new `StudentModuleData` object is created and added to the collection for each student.
```

> Given below are the different steps taken that when the user tries to assign students to a class.

Reviewer: ```suggestion
Given below are the different steps taken that the user assigns students to a module class.
```

Step 1: The user enters the command keyword `assign`, followed by the indices of the students he want to assign classes to, and the names of the classes that he wants to assign to the students. Example: `assign 1 2 3 c/CS1101S`.

> Step 2: The program makes use of the `TAAssistParser` to make sense of the keyword, and determine which parser to use to parse the arguments. In this case, the `AssignCommandParser` will be used.

Reviewer: I think replacing instances of `the program` with `Ta Assist` will match the context better.

Author: I am afraid that might create confusion with the `TaAssist` class within the program. :(

Reviewer: I feel that it's fine as long as we don't make it a code block (i.e. "TA Assist" vs `taAssist`). Anyway I think this can be left later when we standardise everything

Step 3: The `AssignCommandParser` makes sense of the arguments, and creates an `AssignCommand` object with the student indices and the classes to assign them to.

> Step 4: The `AssignCommand` object is executed. The student indices will be used to retrieve the `Student` objects from the list of students captured by the `Model` interface. For each student, the program will create a new `StudentModuleData` object for each class that is not already assigned to the student. The `StudentModuleData` object will only contain the class name and not any session information. The `StudentModuleData` objects created for the student will be added to the student object's collection of `StudentModuleData`.

Reviewer: I think certain phrases like `student indices will be used to...` or `the program will create...` can be rephrased to something like `student indices are used to...` or `the program creates...` to stay consistent.

Step 5: The execution ends and returns a `CommandResult` object containing the success message to be displayed by the GUI to the user.

#### Design considerations:

* Alternative 1 (current choice): Let each student maintain a collection of classes that the student is being assigned to.
  * Pros: Only captures necessary information, and easier to implement. This structure is also easier to capture session information for the students.

>   * Cons: Will be creating multiple `StudentModuleData` objects for a class when multiple students are assigned to the class. This can possibly cause performance issue due to the number of objects created.

Reviewer: ```suggestion
  * Cons: Will be creating multiple `StudentModuleData` objects for a class when multiple students are assigned to the class. This may cause performance issues from the large number of objects created.
```
